src/back/glsl_common.rs

```diff
@@ -0,0 +1,704 @@
+                builder.expressions[*right].write_glsl(module, builder)?
+            ),
+            Expression::CrossProduct(left, right) => format!(
+                "cross({},{})",
+                builder.expressions[*left].write_glsl(module, builder)?,
+                builder.expressions[*right].write_glsl(module, builder)?
+            ),
+            Expression::Derivative { axis, expr } => format!(
+                "{}({})",
+                match axis {
+                    DerivativeAxis::X => "dFdx",
+                    DerivativeAxis::Y => "dFdy",
+                    _ => todo!(),
+                },
+                builder.expressions[*expr].write_glsl(module, builder)?
+            ),
+            Expression::Call { origin, arguments } => format!(
+                "{}({})",
+                match origin {
+                    FunctionOrigin::External(name) => name,
+                    FunctionOrigin::Local(handle) => builder.functions.get(&handle).unwrap(),
+                },
+                arguments
+                    .iter()
+                    .map(|arg| builder.expressions[*arg].write_glsl(module, builder))
+                    .collect::<Result<Vec<_>, _>>()?
+                    .join(","),
+            ),
+        })
+    }
+}
+
+impl Constant {
+    pub(crate) fn write_glsl(
+        &self,
+        module: &Module,
+        builder: &StatementBuilder<'_>,
+    ) -> Result<String, Error> {
+        Ok(match &self.inner {
+            ConstantInner::Sint(int) => int.to_string(),
+            ConstantInner::Uint(int) => int.to_string(),
+            ConstantInner::Float(float) => float.to_string(),
+            ConstantInner::Bool(boolean) => boolean.to_string(),
+            ConstantInner::Composite(components) => format!(
+                "{}({})",
+                match module.types[self.ty].inner {
+                    TypeInner::Vector { size, .. } => format!("vec{}", size as u8,),
+                    TypeInner::Matrix { columns, rows, .. } =>
+                        format!("mat{}x{}", columns as u8, rows as u8,),
+                    TypeInner::Struct { .. } => builder.structs.get(&self.ty).unwrap().0.clone(),
+                    TypeInner::Array { .. } =>
+                        self.ty.write_glsl(builder.types, builder.structs)?,
+                    _ =>
+                        return Err(Error::Custom(format!(
+                            "Cannot build constant of type {}",
+                            self.ty.write_glsl(builder.types, builder.structs)?
+                        ))),
+                },
+                components
+                    .iter()
+                    .map(|component| module.constants[*component].write_glsl(module, builder))
+                    .collect::<Result<Vec<_>, _>>()?
+                    .join(","),
+            ),
+        })
+    }
+}
+
+impl Handle<Type> {
+    pub(crate) fn write_glsl<'a>(
+        &self,
```

Comment:
I find it weird that this function is implemented for a handle. Could we have it as a free function instead, i.e. `write_type`?